include/lsst/pex/exceptions/python/Exception.h

```diff
@@ -0,0 +1,84 @@
+// -*- LSST-C++ -*-
+/* 
+ * LSST Data Management System
+ * Copyright 2008-2016  AURA/LSST.
+ * 
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the LSST License Statement and 
+ * the GNU General Public License along with this program.  If not, 
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_PEX_EXCEPTIONS_PYTHON_EXCEPTION_H
+#define LSST_PEX_EXCEPTIONS_PYTHON_EXCEPTION_H
+
+#include "lsst/pex/exceptions/Exception.h"
+
+#include <pybind11/pybind11.h>
+
+#include <ostream>
+#include <string>
+
+namespace lsst {
+namespace pex {
+namespace exceptions {
+namespace python {
+
+/** Helper function for pybind11, used to define new types of exceptions.
+  * \param[in] mod Module to insert the exception into.
+  * \param[in] name Name of the exception in the module.
+  * \param[in] base Python name of base class (from pex exceptions).
+  */
+template <typename T>
+void declare_exception(pybind11::module &mod, const std::string & name, const std::string & base) {
+    using lsst::pex::exceptions::Exception;
+
+    // Wrap T as a new Python exception type with *name* and add it to the module
+    //
+    // Note that all created C++ wrapped type derive from Exception here.
+    // It is only in the pure Python wrapper layer that they get embedded in
+    // a subclass of the requested base.
+    pybind11::class_<T> cls(mod, name.c_str(), pybind11::base<Exception>());
+    cls.def(pybind11::init<char const*, int, char const*, std::string const &>())
+        .def(pybind11::init<std::string const &>());
+    
+    // Declare T wrapped by cls as a pex exception and register it
+    // this relies on the pure Python function "declare" defined in "wrappers"
+    // to create a corresponding Python type derived from Python standard Exception
+    PyObject* exceptions = PyImport_ImportModule("lsst.pex.exceptions.wrappers");
+    if (!exceptions) {
+        throw LSST_EXCEPT(RuntimeError, "import failed");
+    }   
+
+    PyObject* declare = PyObject_GetAttrString(exceptions, "declare");
+    if (!declare) {
+        throw LSST_EXCEPT(RuntimeError, "could not get declare function from Python");
+    }   
```

Comment:
If this check fails but the one above did not, `exceptions` will leak.  I think you can use pybind11's `handle` class as a smart pointer for `PyObject` to avoid this problem with RAII.